hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -505,7 +505,16 @@ func (r *NodePoolReconciler) reconcile(ctx context.Context, hcluster *hyperv1.Ho
 			return reconcile.Result{}, fmt.Errorf("failed to get token Secret: %w", err)
 		}
 		if err == nil {
-			if err := setExpirationTimestampOnToken(ctx, r.Client, tokenSecret); err != nil && !apierrors.IsNotFound(err) {
+			_, err := r.CreateOrUpdate(ctx, r.Client, tokenSecret, func() error {
+				// this should be a reasonable value to allow all in flight provisions to complete.
+				timeUntilExpiry := 2 * time.Hour
+				if tokenSecret.Annotations == nil {
+					tokenSecret.Annotations = map[string]string{}
+				}
+				tokenSecret.Annotations[hyperv1.IgnitionServerTokenExpirationTimestampAnnotation] = time.Now().Add(timeUntilExpiry).Format(time.RFC3339)
+				return nil
+			})
+			if err != nil && !apierrors.IsNotFound(err) {
```

Comment:
The `NotFound` is impossible now because it will get created if it doesn't exist, is that what we want? cc @enxebre

Reply:
right, for this particular case we just don't want to create or update something, we just want to update it if it exists so I guess current code should remain and an improvement would be to avoid the update call if the annotation is already in it. Also we might want to consolidate on patching instead of updating?

Reply:
One exception in the CPO is the finalizer handling, which is handled before all other reconciliation and uses patch. Otherwise there is expected to be only 1 call to CreateOrUpdate per resource per reconcile (including HostedControlPlane itself). I think.

Reply:
>Contrast with CPO which was designed internally with the controller-runtime CreateOrUpdate pattern from the beginning, where there is a single function to handle create/update for every discrete resource.

The NodePool controller uses CreateOrUpdate pattern from the beginning as well where the semantic applies i.e your intent is to create or update.

mm I can't think how refactoring can help here, the intent in this particular case will still be "update if exists and a condition is true, do nothing if it doesn't".

Reply:
Looked more closely at the existing nodepool controller, here are my thoughts:

1. The finalizer management could use patch like you mentioned and that would be consistent w/ the CPO pattern to special-case finalization
2. The [reconcile logic for the nodepool itself](https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/nodepool/nodepool_controller.go#L193) contains code that mutates many other resources and then does a patch of the nodepool afterwards outside the reconcile function; what I expected instead is that nodepool reconcile logic is a single function that mutates the nodepool and no other resources
3. TokenSecret is reconciled in two different paths ([1](https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/nodepool/nodepool_controller.go#L501), [2](https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/nodepool/nodepool_controller.go#L527)). These could be consolidated (by using `isUpdatingVersion` and `isUpdatingConfig` as inputs to the reconcile function)

Reply:
> what I expected instead is that nodepool reconcile logic is a single function that mutates the nodepool and no other resources

Not sure I follow this one, nodepool controller for sure needs to reconcile other resources based on the NodePool spec, for those the pattern used is createOrUpdate where it makes sense. This is the same than for any other controller...
Now the pattern that wraps the whole reconcile into a patch for the NodePool (mainly for annotation and status) that could be changed and made inline if we wanted to.

>TokenSecret is reconciled in two different paths ([1](https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/nodepool/nodepool_controller.go#L501), [2](https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/nodepool/nodepool_controller.go#L527)). These could be consolidated (by using isUpdatingVersion and isUpdatingConfig as inputs to the reconcile function)

true, we can do that.

Reply:
>Not sure I follow this one, nodepool controller for sure needs to reconcile other resources based on the NodePool spec, for those the pattern used is createOrUpdate where it makes sense. 

The code that is intended to create or update the nodepool resource itself contains interspersed code for creating and updating other resources (secrets, ibmpowervsimage, machinetemplates, etc.). What I'm saying is that the pattern I expected was for a given function to deal with creating and updating a single resource. So in this case, I would expect the nodepool function to only create or update a nodepool. And if any of those other resources are inputs to the nodepool function, those resources would be reconciled first, and then passed as inputs to the nodepool function.

>This is the same than for any other controller...

Not exactly. In the CPO you'll notice that the hostedcontrolplane controller mutates status of HCP, and then reconciles all other resources and does not mutate the HCP as part of that reconciliation. HostedCluster controller follows that pattern as well. In NodePool controller the nodepool reconcile function mutates status (and at least one non-status item with ownerreferences), plus all the other resources mentioned before, and then finally patches the whole resource.

I'm not sure any one of them is "correct", but I am just pointing out that NodePool controller seems to be subtly different than HostedControlPlane/HostedCluster with regards to the reconcile pattern of the core resource.